pkg/frontend/frontend_test.go

```diff
@@ -99,7 +101,139 @@ func TestFrontend_RequestHostHeaderWhenDownstreamURLIsConfigured(t *testing.T) {
+		serverClusterValidation clusterutil.ServerClusterValidationConfig
+		clientClusterValidation clusterutil.ClusterValidationConfig
+		expectedMetrics         string
+		serverCluster           string
+		enabled                 bool
+		softValidation          bool
+		clientCluster           string
+		expectedError           string
+		expectedStatusCode      int
+	}{
+		"if server and client have equal cluster validation labels and cluster validation is disabled no error is returned": {
+			clientCluster:      "cluster",
+			serverCluster:      "cluster",
+			enabled:            false,
+			expectedStatusCode: 200,
+		},
+		"if server and client have different cluster validation labels and cluster validation is disabled no error is returned": {
+			clientCluster:      "client-cluster",
+			serverCluster:      "server-cluster",
+			enabled:            false,
+			expectedStatusCode: 200,
+		},
+		"if server and client have equal cluster validation labels and cluster validation is enabled no error is returned": {
+			clientCluster:      "cluster",
+			serverCluster:      "cluster",
+			enabled:            true,
+			expectedStatusCode: 200,
+		},
+		"if server and client have different cluster validation labels and soft cluster validation is enabled no error is returned": {
+			clientCluster:      "client-cluster",
+			serverCluster:      "server-cluster",
+			enabled:            true,
+			softValidation:     true,
+			expectedStatusCode: 200,
+		},
+		"if server and client have different cluster validation labels and cluster validation is enabled an error is returned": {
+			clientCluster:      "client-cluster",
+			serverCluster:      "server-cluster",
+			enabled:            true,
+			softValidation:     false,
+			expectedStatusCode: 500,
+			expectedError:      `request rejected by the server: rejected request with wrong cluster validation label "client-cluster" - it should be "server-cluster"`,
+			expectedMetrics: `
+				# HELP cortex_client_request_invalid_cluster_validation_labels_total Number of requests with invalid cluster validation label.
+        	    # TYPE cortex_client_request_invalid_cluster_validation_labels_total counter
+        	    cortex_client_request_invalid_cluster_validation_labels_total{client="query-frontend",method="/api/v1/query_range",protocol="http"} 1
+			`,
+		},
+		"if client has no cluster validation label and soft cluster validation is enabled no error is returned": {
+			clientCluster:      "",
+			serverCluster:      "server-cluster",
+			enabled:            true,
+			softValidation:     true,
+			expectedStatusCode: 200,
+		},
+		"if client has no cluster validation label and cluster validation is enabled an error is returned": {
+			clientCluster:      "",
+			serverCluster:      "server-cluster",
+			enabled:            true,
+			softValidation:     false,
+			expectedStatusCode: 511,
+			expectedError:      "rejected request with empty cluster validation label - it should be \\\"server-cluster\\\"",
```

Comment:
[Nit] Could you write this as a raw string literal instead, so we can avoid so many backslashes for readability?